library/core/src/ptr/non_null.rs

```diff
@@ -1770,3 +1779,30 @@ impl<T: ?Sized> From<&T> for NonNull<T> {
         unsafe { NonNull { pointer: reference as *const T } }
     }
 }
+
+//#[unstable(feature="kani", issue="none")]
+mod verify {
+    use super::*;
+
+    // pub const unsafe fn new_unchecked(ptr: *mut T) -> Self
+    #[kani::proof_for_contract(NonNull::new_unchecked)]
+    pub fn non_null_check_new_unchecked() {
+        let mut x : i32 = kani::any();
+        let xptr = &mut x;
+        unsafe {
+            let _ = NonNull::new_unchecked(xptr as *mut i32);
+        }
+    }
+
+    // pub const unsafe fn new(ptr: *mut T) -> Option<Self>
+    #[kani::proof_for_contract(NonNull::new)]
+    pub fn non_null_check_new() {
+        let mut x : i32 = kani::any();
+        let xptr = &mut x;
+        unsafe {
+            let nonnull_ptr = NonNull::new(xptr as *mut i32);
+            let null_ptr = NonNull::<u32>::new(null_mut());
```

Comment:
You could perhaps simplify this by making a single call to `new`, e.g.:
```rust
let maybe_null_ptr =  if kani::any() { xptr as *mut i32 } else { null_mut() };
let _ = NonNull::new(maybe_null_ptr);
```

Reply:
Is it because `kani::any()` could potentially return a zero value which will make the if statement false?

Reply:
Yes, `kani::any()` in this context will return true or false, which will effectively make the pointer either NULL or non-NULL, thus covering the full space of inputs for the function under verification.